README
rmmq version 1.01
=======================

NAME:

Range Minimum Maximum Query


SYNOPSIS:

    #include<vector>
    #include<rmmq.hpp>
 
    vector<int|long|unsigned|double> vec {1,5,23,7,8,3,12,5,3,44,56};

/* sparse table */
    /* Construct */
    ST<int|long|unsigned|double> sptab(vec);
    /* OR */
    ST<int|long|unsigned|double> sptab;
    sptab.make(vec)
    
    /* Explicite Destructor */
    sptab.destroy();



    
/* rmmq */
    /* Construct */
    RMMQ<int|long|unsigned|double> rmmq(vec);
    /* OR */
    RMMQ<int|long|unsigned|double> rmmq;
    rmmq.make(vec)

    /* Explicite Destructor */
    rmmq.destroy();
  




    /* Range Minimum Query for (5,3) */
    /* return value */
    rmmq.MinVal(5-1,3-1); // returns 7  (-1 is for vector index positions since indexing starts from 0)

    /* return position */
    rmmq.MinPos(5-1,3-1); // returns 3 since indexing starts from 0 (-1 is for vector index positions since indexing starts from 0)

    /* Range Maximum Query for (3,5) */
    /* return value */
    rmmq.MaxVal(3-1,5-1); // returns 23  (-1 is for vector index positions since indexing starts from 0)

    /* return position in array */
    rmmq.MaxPos(3-1,5-1); // returns 2 since indexing starts from 0 (-1 is for vector index positions since indexing starts from 0)


DESCRIPTION:

Given an array of numbers (a) and two position i,j\in [1,|a|], program computes the minimum and/or maximum position and/or value within the interval specified by the two numbers (i and j).



CONSTRUCTION AND RUNTIME COMPLEXITY:

Complexity for n = |a| (number of elements in a given array):

    Construction:  O(n log n)
    
    Retrieve:    O(1)


ACKNOWLEDGMENTS:
     
     <http://www.topcoder.com/tc?module=Static&d1=tutorials&d2=lowestCommonAncestor>

AUTHOR:

Robert Bakaric <rbakaric@irb.hr>, <bakaric@evolbio.mpg.de>

COPYRIGHT AND LICENSE:

 * Copyright 2015 Robert Bakaric
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.



ChangeLog
27.03.15	Created! 
19.02.16	Correcting errors in make() and destroy() functions reported by  Giacomo Donati.


src/include/rmmq.hpp
/*
 * rmmq.hpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */


#include<iostream>
#include<vector>
#include<string>
#include<cmath>
#include<algorithm>

using namespace std;

template<typename TInt>
class ST{
  protected:
  
  vector<vector<TInt>> MinSTable;
  vector<vector<TInt>> MaxSTable;  
  
  public:

/* Constructor */
  ST();
  ST(const vector<TInt>& rvec);  

/* Destructor */
  ~ST();


/* Builder */
  void MakeST(const vector<TInt>& rvec);
  
  void make(const vector<TInt>& rvec);
  void destroy();

/* Getter -- not finished */ 
  TInt Get(const TInt x, const TInt y);
  
};


/* Constructor */
template <typename TInt>
ST<TInt>::ST(){}

template <typename TInt>
ST<TInt>::ST(const vector<TInt>& rvec){
  MakeST(rvec);
}

template <typename TInt>
void ST<TInt>::make(const vector<TInt>& rvec){
  MakeST(rvec);
}


/* Destructor */
template <typename TInt>
ST<TInt>::~ST(){
  vector<vector<TInt>>().swap(MinSTable);
  vector<vector<TInt>>().swap(MaxSTable);
}

template <typename TInt>
void ST<TInt>::destroy(){
  vector<vector<TInt>>().swap(MinSTable);
  vector<vector<TInt>>().swap(MaxSTable);
}



/* Functions Make ST */
template <typename TInt>
void ST<TInt>::MakeST(const vector<TInt>& rvec){

  TInt x = log2(rvec.size()) + 1;
  TInt p = 0;

  vector<TInt> vec(x);

  MinSTable.resize(rvec.size()+1, vec);
  MaxSTable.resize(rvec.size()+1, vec);

  for(TInt i = 0; i < rvec.size(); i++){
    MinSTable[i][0] = i;
    MaxSTable[i][0] = i;
  }
  
  for(TInt i = 1; i< x; i++){
    p=(1 << (i - 1));
    for(TInt j=0; j+(1<<i)-1 < rvec.size(); j++){
      MinSTable[j][i] = (rvec[MinSTable[j][i-1]] < rvec[MinSTable[j+p][i-1]]) ? (MinSTable[j][i-1]) : (MinSTable[j+p][i-1]);
      MaxSTable[j][i] = (rvec[MaxSTable[j][i-1]] > rvec[MaxSTable[j+p][i-1]]) ? (MaxSTable[j][i-1]) : (MaxSTable[j+p][i-1]);
    }
  }
}





template<typename TInt>
class RMMQ : protected ST<TInt>{
  
  vector <TInt> range;
/* Functions */
  
  public:
/* Constructor */
  RMMQ();
  RMMQ(const vector <TInt>& vec);

/* Destructor */
  ~RMMQ();

/* Queries */

  TInt MinVal(const TInt, const TInt);
  TInt MaxVal(const TInt, const TInt);
  TInt MinPos(const TInt, const TInt);
  TInt MaxPos(const TInt, const TInt); 

  void make(const vector<TInt>& vec);
  void destroy();
};

/* Constructors */

template <typename TInt>
RMMQ<TInt>::RMMQ():ST<TInt>(){}

template <typename TInt>
RMMQ<TInt>::RMMQ(const vector<TInt>& vec) : range(vec), ST<TInt> (vec){}

/* Destructor */
template <typename TInt>
RMMQ<TInt>::~RMMQ(){
  vector<TInt>().swap(range);
}


template <typename TInt>
void RMMQ<TInt>::destroy(){
  vector<TInt>().swap(range);
  ST<TInt>::destroy();
}


/* Functions */
template <typename TInt>
void RMMQ<TInt>::make(const vector<TInt>& vec){
 range = vec;
 ST<TInt>::make(vec);
}

/* Queries */
template <typename TInt>
TInt RMMQ<TInt>::MinVal(const TInt x, const TInt y){
 
  TInt mn = min(x,y);
  TInt mx = max(x,y);

  TInt o = log2(mx - mn + 1);
  TInt p = (1<<o);
  return (range[this->MinSTable[mn][o]] < range[this->MinSTable[mx - p + 1][o]]) ? (range[this->MinSTable[mn][o]]) : (range[this->MinSTable[mx - p + 1][o]]);
}

  


template <typename TInt>
TInt RMMQ<TInt>::MaxVal(const TInt x, const TInt y){

  TInt mn = min(x,y);
  TInt mx = max(x,y);

  TInt o = log2(mx - mn + 1);
  TInt p = (1<<o);
  return (range[this->MaxSTable[mn][o]] > range[this->MaxSTable[mx - p + 1][o]]) ? (range[this->MaxSTable[mn][o]]) : (range[this->MaxSTable[mx - p + 1][o]]);
}


template <typename TInt>
TInt RMMQ<TInt>::MinPos(const TInt x, const TInt y){
 
  TInt mn = min(x,y);
  TInt mx = max(x,y);
  
  TInt o = log2(mx - mn + 1);
  TInt p = (1<<o);
  return (range[this->MinSTable[mn][o]] < range[this->MinSTable[mx - p + 1][o]]) ? (this->MinSTable[mn][o]) : (this->MinSTable[mx - p + 1][o]);
}




template <typename TInt>
TInt RMMQ<TInt>::MaxPos(const TInt x, const TInt y){
  
  TInt mn = min(x,y);
  TInt mx = max(x,y);

  TInt o = log2(mx - mn + 1);
  TInt p = (1<<o);
  return (range[this->MaxSTable[mn][o]] > range[this->MaxSTable[mx - p + 1][o]]) ? (this->MaxSTable[mn][o]) : (this->MaxSTable[mx - p + 1][o]);
}



src/apps/rmmq.cpp
/*
 * rmmq.cpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include<iostream>
#include<fstream>
#include<vector>
#include<string>

#include "boost/variant.hpp"
#include <boost/program_options.hpp>

#include<rmmq.hpp>


namespace po = boost::program_options;
using namespace std;



void PrintAcknowledgements(){

cout << "**************************************************************************************"<<"\n";
cout <<"                         rmmq - Range minimum/maximum query                                "<<"\n";
cout <<"                                          by                                           "<<"\n";
cout <<"                                    Robert Bakaric                                     "<<"\n\n";
cout <<"CONTACT:                                                                               "<<"\n";
cout <<" Code written and maintained by Robert Bakaric,                                        "<<"\n";
cout <<" email: rbakaric@irb.hr , bakaric@evolbio.mpg.de                                       "<<"\n\n";
cout <<"ACKNOWLEDGEMENT:                                                                       "<<"\n";
cout <<" http://www.topcoder.com/tc?module=Static&d1=tutorials&d2=lowestCommonAncestor         "<<"\n\n";
cout <<"LICENSE:                                                                               "<<"\n";
cout <<" The program is distributed under the GNU General Public License. You should have      "<< "\n";
cout <<" received a copy of the licence together  with this software. If not, see              "<< "\n";
cout <<" http://www.gnu.org/licenses/                                                          "<< "\n";
cout <<"***************************************************************************************"<< "\n\n\n";
}


template <typename INT, typename CHARA>
po::variables_map SetOptions(INT& argc, CHARA& argv){

    try {
        int opt;
        string version = "1.01";
        po::options_description desc("Allowed options");
        desc.add_options()
            ("help,h", "produce help message")
            ("version,v", "print version information")
            ("input-file,i", po::value< string >(), "input file")
        ;

        po::positional_options_description p;
        p.add("input-file,i", -1);
        
        po::variables_map vm;
        po::store(po::command_line_parser(argc, argv).
                  options(desc).positional(p).run(), vm);
        po::notify(vm);
    
        if (vm.count("help")) {
            cout << "Usage: ./program [options]\n\n";
            PrintAcknowledgements();
            cout << desc;
            exit(0);
        }else if (vm.count("version")) {
            cout << "Program version:  " << version << "\n";
            exit(0);
        }
        if (!vm.count("input-file")){
            cout << "Input file is not defined \n";
            exit(0);
        }
        return vm;    
    }
    catch(std::exception& e)
    {
        cout << e.what() << "\n";
        exit(0);
    }    
}

void ReadRange(const string & file, vector<int>& vec){
  
  fstream fs;
  fs.open (file.c_str(), ios::in);
  if ( !fs.is_open()){
    cerr << "Cannot open file: "<<file <<"! Line:" << __LINE__<< endl;
    exit(-1);
  }

  int i = 0; 
  while( !fs.eof()) {
    if(vec.size() == i)
       vec.resize((vec.size()+1)*2);
    fs >> vec[i++];
    
  }  
 
  vec.resize(i-1);
  fs.close();
}




int main (int argc, char** argv){
  
  
  
  po::variables_map arg;
  arg = SetOptions(argc,argv);

  string file =  arg["input-file"].as<string>();

   vector<int> vec;
   
   ReadRange(file, vec);
/*Print array */
   cout << "Positions:";
   for(int i = 0; i< vec.size(); i++){
      cout << "\t[" << i+1 <<"]";
   }
   cout << endl;
   cout << "Values:   ";  
   for(int i = 0; i< vec.size(); i++){
      cout << "\t" << vec[i] ;
   }
   cout << endl;
/* Build ST */
   RMMQ<int> rmmq(vec);

/* Process Queries */
  int in = 0;
  int x=0, y=0;
  
  while(1){
    cout << "Note: Ctrl-c to quit\n";
    cout << "Start: ";
    cin >> x;
    cout << "Stop: ";
    cin >> y;
    
    if(x<=vec.size() && y<=vec.size() && x>=0 && y>=0){ 
      cout << "\nmin("<< x<<","<<y<<"): val(" << rmmq.MinVal(x-1,y-1) <<") pos(" << rmmq.MinPos(x-1,y-1)+1 << ")"<< endl;
      cout << "max("<< x<<","<<y<<"): val(" << rmmq.MaxVal(x-1,y-1) <<") pos(" << rmmq.MaxPos(x-1,y-1)+1 << ")\n"<< endl;
    }else
      cout << "\nValues are outside the given interval\n";
  }
      
  
  return 0;
}
